### PR TITLE
feat: Add a content field and element to raw string literals.

### PR DESCRIPTION
**Pull request by Natr1x**
This makes it possible to use language injection for this content.
Which is useful since this is something that raw string literals are often used for.

A query like the following could for example be used to provide sql highlighting when using sqlx:
```clojure
; extends
((macro_invocation
   macro: (scoped_identifier
            path: (identifier) @macropath (#eq? @macropath "sqlx")
            name: (identifier) @macroname)
   (token_tree
     (raw_string_literal
       content: (_) @injection.content)))
 (#set! injection.language "sql")
)
``` 

__Note: I have not ran `tree-sitter-cli generate` in order to keep the commit smaller for review purposes.__

If you are interested in merging this but want me to change or add something (or run that code generation) then I am willing to do so.

### CODE REVIEW

File: src/scanner.c [hunk 1]
```diff
@@ -1,18 +1,46 @@
...
+  RAW_STRING_LITERAL_CLOSE,
   FLOAT_LITERAL,
   BLOCK_COMMENT,
 };
 
-void *tree_sitter_rust_external_scanner_create() { return NULL; }
-void tree_sitter_rust_external_scanner_destroy(void *p) {}
+struct ScannerState {
+  unsigned opening_hash_count;
+};
```
Comment (amaanq) [inline src/scanner.c:16]: A struct for one field is a bit overkill, let's just allocate a uint32_t instead (I'd rather not use unsigned)

File: src/scanner.c [hunk 2]
```diff
@@ -1,18 +1,46 @@
+#include <string.h>
 #include <tree_sitter/parser.h>
 #include <wctype.h>
 
 enum TokenType {
   STRING_CONTENT,
-  RAW_STRING_LITERAL,
+  RAW_STRING_LITERAL_OPEN,
+  RAW_STRING_LITERAL_CONTENT,
+  RAW_STRING_LITERAL_CLOSE,
```
Comment (amaanq) [inline src/scanner.c:9]: Maybe call the delimiters START/END? Unless the official grammar calls them that

File: src/scanner.c [hunk 3]
```diff
@@ -42,42 +72,76 @@ bool tree_sitter_rust_external_scanner_scan(void *payload, TSLexer *lexer,
...
 
     if (lexer->lookahead != '"') return false;
+
     advance(lexer);
+    return true;
+  }
+
+  if (
+    valid_symbols[RAW_STRING_LITERAL_CONTENT]
+  ) {
```
Comment (amaanq) [inline src/scanner.c:98]: Slight nitpicks but don't be overzealous with newlines/spacing stuff out, it's a bit cleaner if it's one condition to just have it on the same line, and same for the newlines below with the `}`s

**Review by amaanq — CHANGES_REQUESTED** [review]

File: src/scanner.c (hunk 1 above)
Comment (Natr1x) [reply src/scanner.c]: Are you sure about this? Personally I feel keeping it in a struct makes it explicitly clear that this is the only state we keep track of.

File: src/scanner.c (hunk 3 above)
Comment (Natr1x) [reply src/scanner.c]: Fixed

File: src/scanner.c (hunk 1 above)
Comment (amaanq) [reply src/scanner.c]: alright yeah, if we do need to add more fields later on for whatever reason it'd be easier. Can we declare it as a typedef, use uint32_t, and call it just Scanner though?
```c
typedef struct {
    uint32_t opening_hash_count;
} Scanner
```

File: src/scanner.c (hunk 2 above)
Comment (amaanq) [reply src/scanner.c]: Closest I could find was https://github.com/rust-lang/wg-grammar/ which just calls it hashes, doesn't matter

File: src/scanner.c (hunk 1 above)
Comment (Natr1x) [reply src/scanner.c]: I am not a big fan of anonymous structs, would f6caf4ae21733980406e8982c4349ea44181731f work as a compromise?

If not I am okay with removing the name. It's mostly for the sake of things like lsp hover information.

**Comment by mkatychev** [comment]
@Natr1x what else needs to get done for this PR to move out of draft. This would be extremely useful for sql injections and would remove `#offset!` guessworkd.

**Comment by amaanq** [comment]
reworked it in #220

**Pull request closed without merging**